### PR TITLE
Add Make.win32 -fcommon,  define u64int, sync libsec, libmp, libauthsrv with Plan 9, define usize, add chacha20, libsec: add mising \

### PR DESCRIPTION
**Pull request by adriangrigore**

### CODE REVIEW

**Comment by 0intro** [comment]
Thanks. I'm looking at it.

**Comment by adriangrigore** [comment]
I removed `#pragma`s at start of  header files, maybe it's a good idea to leave them there despite the warnings. Proper solution I think would be to `#ifdef` them: https://github.com/9fans/drawterm/blob/main/include/authsrv.h but that would mean patching on Plan 9 side also.

**Comment by adriangrigore** [comment]
Oh, I see there are more changes in this repo vs Plan 9 versions(not that many) I just copied the libs.

I could merge and force push I think.

**Comment by 0intro** [comment]
Overall, it looks good. Thank your for working on this.

Could you rename `libsec: sync with Plan9` to `libsec: sync with Plan 9`?

**Comment by adriangrigore** [comment]
Ok, changed libsec commit message, left `#pragma`s in place and added `-Wno-unknown-pragmas` to `Make.win32`.

**Comment by 0intro** [comment]
I think it's probably better to add `#ifdef PLAN9` around pragma, like it was done for the other header files (`include/draw.h` and so on).

**Comment by adriangrigore** [comment]
I took the flag from here https://github.com/golang/go/blob/ceb95ea6aef52c1fb472d3539c6ef68670778b5b/src/cmd/dist/build.c#L482, not sure if it applies to Plan 9, and I think it's preferable to have identical copies in drawterm and Plan 9. If I were to add `#ifdef`s, I'd probably add them to the Plan 9 versions also to keep them precisely in sync. That's why I opted for ` -Wno-unknown-pragmas`.

**Comment by adriangrigore** [comment]
plan9port removes them completely https://github.com/9fans/plan9port/blob/master/include/mp.h#L137 or comments them out https://github.com/9fans/plan9port/blob/master/include/mp.h#L151, I understand your argument of `#idef`ing them to keep consistency with the current codebase, I would agree with that for consistency purposes.

**Comment by adriangrigore** [comment]
Shall we go with consistency then?

**Comment by adriangrigore** [comment]
Although, not all them were `#ifdef`-ed, some just had their `#pragma`s removed https://github.com/9fans/drawterm/blob/main/include/libsec.h,  https://github.com/9fans/drawterm/blob/main/include/mp.h.

**Comment by adriangrigore** [comment]
If you prefer `#ifdefs`, 9legacy solution would probably be create some patches to `ifdef` them in Plan 9 too I think. To keep them perfectly in sync, wih no changes to the Plan 9 versions, ` -Wno-unknown-pragmas` would probably be the solution.

**Comment by adriangrigore** [comment]
Or just live with the warnings.

**Comment by 0intro** [comment]
The pragma are meant to be interpreted by the Plan 9 C compiler only. I'm not very comfortable keeping them, risking to be handled by the GCC preprocessor, which could lead to side effects.

Either deleting them or `#ifdef` sounds good to me.

**Comment by adriangrigore** [comment]
> The pragma are meant to be interpreted by the Plan 9 C compiler only. I'm not very comfortable keeping them, risking to be handled by the GCC preprocessor, which could lead to side effects.
> 
> Either deleting them or `#ifdef` sounds good to me.

Ok, I'll probably `#ifdef` them then.

**Comment by adriangrigore** [comment]
Ok, `#ifdef`-ed, not sure if this is ok https://github.com/9fans/drawterm/pull/16/commits/0f4e314e5b7bd8d1327d758eded16ca578a5d761#diff-e9399bd85711a4c93d1a01968311639e3355a8b3dce252f026d44cd7650ae913R130, if ok, I think you can merge.

**Comment by adriangrigore** [comment]
Hmm... this might require merging https://github.com/9fans/drawterm/commit/93a43be0ea85aca12826c9d53aec0ad2a0278198

**Comment by adriangrigore** [comment]
I don't understand which (if any of the patches) to `libmp`, `libsec`, `libauthsrv` from this repo have been merged in Plan 9 libs, if not, we'll have to merge them.

**Comment by adriangrigore** [comment]
> Hmm... this might require merging [93a43be](https://github.com/9fans/drawterm/commit/93a43be0ea85aca12826c9d53aec0ad2a0278198)

I see block files are done in assembly, I added the C versions https://github.com/9fans/drawterm/pull/16/commits/15a79f4db957e04ddb129723d0bc7767da6f13c5#diff-ede5614f90d7b8d84cda78003b911a92e627073420f71965b843c0c4bbab4921R50 for `sha2block64` and `sha2block128`.

**Comment by adriangrigore** [comment]
https://github.com/9fans/drawterm/commit/1c8b4992284fe1cbccb54cd213f274c00c3ac6d1#diff-60bc2862e39ce285a55d63a7f349f7eb7e5724d0f1bc0baa0cc18a28e0ff5936 this affects `libmp` (not merged).

**Comment by adriangrigore** [comment]
https://github.com/9fans/drawterm/commit/6fb85848001386a1acf2bc47605b797ce21c48a7 and https://github.com/9fans/drawterm/commit/e10503f0194770f2f5350775ada3570ee24f5944 look merged.

**Comment by adriangrigore** [comment]
https://github.com/9fans/drawterm/commit/1125f094ce2a6854ef568b8c62602ce07c6a2935#diff-ba43ca4048f51025813c140fdcc7f53f77817dd416f1aa445900c524e5090b86L1 affects `libmp`, not merged.

**Comment by adriangrigore** [comment]
https://github.com/9fans/drawterm/commit/9320b94c17a5009a7b50918d38842e268749f752#diff-7cbc4f84d9558c6a8a3236a3e94b85584d7d80805284758fb4ad0abe11836844L61 looks meged

**Comment by 0intro** [comment]
> I don't understand which (if any of the patches) to `libmp`, `libsec`, `libauthsrv` from this repo have been merged in Plan 9 libs, if not, we'll have to merge them.

Merging is always complicated because you have to compare two pieces of code that evolved independently.

**Comment by adriangrigore** [comment]
> > I don't understand which (if any of the patches) to `libmp`, `libsec`, `libauthsrv` from this repo have been merged in Plan 9 libs, if not, we'll have to merge them.
> 
> Merging is always complicated because you have to compare two pieces of code that evolved independently.

I assume they started from the same sources, and I assume the idea was for them to be sync-ed, if so, they don't seem to have a lot of changes between them. Please take a look at `libsec`, `libmp` and `libauthsrv` commit history in this repo, there's not a lof of changes. Doesn't seem hard to merge. Changes are minor.

**Comment by 0intro** [comment]
> [1125f09#diff-ba43ca4048f51025813c140fdcc7f53f77817dd416f1aa445900c524e5090b86L1](https://github.com/9fans/drawterm/commit/1125f094ce2a6854ef568b8c62602ce07c6a2935#diff-ba43ca4048f51025813c140fdcc7f53f77817dd416f1aa445900c524e5090b86L1) affects `libmp`, not merged.

For that kind of changes, you have to re-apply the changes Russ did on your `libmp: sync with Plan 9` commit to be sure there are no regression.

**Comment by adriangrigore** [comment]
> > [1125f09#diff-ba43ca4048f51025813c140fdcc7f53f77817dd416f1aa445900c524e5090b86L1](https://github.com/9fans/drawterm/commit/1125f094ce2a6854ef568b8c62602ce07c6a2935#diff-ba43ca4048f51025813c140fdcc7f53f77817dd416f1aa445900c524e5090b86L1) affects `libmp`, not merged.
> 
> For that kind of changes, you have to re-apply the changes Russ did on your `libmp: sync with Plan 9` commit to be sure there are no regression.

Yup, exactly! Doesn't seem hard! :)

**Comment by adriangrigore** [comment]
Ok ,merged, patches for 9legacy if interested. 

`libmp-gcc-warnings-drawterm.diff`
```
diff --git a/libmp/strtomp.c b/libmp/strtomp.c
index 9c64c33..26e3751 100644
--- a/libmp/strtomp.c
+++ b/libmp/strtomp.c
@@ -32,13 +32,13 @@ init(void)
 	memset(tab.t10, INVAL, sizeof(tab.t10));
 
 	for(p = set64; *p; p++)
-		tab.t64[*p] = p-set64;
+		tab.t64[(uchar)*p] = p-set64;
 	for(p = set32; *p; p++)
-		tab.t32[*p] = p-set32;
+		tab.t32[(uchar)*p] = p-set32;
 	for(p = set16; *p; p++)
-		tab.t16[*p] = (p-set16)%16;
+		tab.t16[(uchar)*p] = (p-set16)%16;
 	for(p = set10; *p; p++)
-		tab.t10[*p] = (p-set10);
+		tab.t10[(uchar)*p] = (p-set10);
 
 	tab.inited = 1;
 }
```
`libmp-sun-fixes-drawterm.diff`
```
diff --git a/libmp/dat.h b/libmp/dat.h
index 7c834ac..8542909 100644
--- a/libmp/dat.h
+++ b/libmp/dat.h
@@ -1,4 +1,4 @@
-#define	mpdighi  (mpdigit)(1<<(Dbits-1))
+#define	mpdighi  (mpdigit)(1U<<(Dbits-1))
 #define DIGITS(x) ((Dbits - 1 + (x))/Dbits)
 
 // for converting between int's and mpint's
```

Now, I really don't want to write `sha2block64` and `sha2block128` in assembler now. :D

**Comment by adriangrigore** [comment]
> Ok, `#ifdef`-ed, not sure if this is ok [0f4e314#diff-e9399bd85711a4c93d1a01968311639e3355a8b3dce252f026d44cd7650ae913R130](https://github.com/9fans/drawterm/commit/0f4e314e5b7bd8d1327d758eded16ca578a5d761#diff-e9399bd85711a4c93d1a01968311639e3355a8b3dce252f026d44cd7650ae913R130), if ok, I think you can merge.

Please review this also, if I didn't miss anything when merging repo changes (I don't think so), it's fine to merge now, except for the asm version of `sha2block64` and `sha2block128` which I don't think affects us at all.

**Comment by 0intro** [comment]
Thanks. I'll will do a final review.

**Comment by adriangrigore** [comment]
Perhaps you might consider merging to a separate branch for now until everything is complete?

File: libsec/Makefile [hunk 1]
```diff
@@ -51,7 +51,7 @@ OFILES=\
 	sha2block64.$O\
 	sha2_128.$O\
 	sha2block128.$O\
-	smallprimes.$O
+	smallprimes.$O\
```
Comment (0intro) [inline libsec/Makefile:54]: This is not necessary, but it doesn't harm.

**Review by 0intro — APPROVED** [review]
I think your changes are ready to be merged on the `main` branch.
You will be able to open new PRs for the remaining changes.

File: libsec/Makefile (hunk 1 above)
Comment (adriangrigore) [reply libsec/Makefile:54]: I guess I preferred single line diffs as I have to add more stuff there. Your call!

**Comment by 0intro** [comment]
I'm ready to merge into `main`. Let me know if it's good for you.

**Comment by adriangrigore** [comment]
Ok to merge!

**Comment by adriangrigore** [comment]
I think this also closes #12

**Pull request merged**